### DISCUSSION TITLE
Enabling Determinism

### DISCUSSION THREAD

**Discussion by torky** [godot-jolt/godot-jolt#890] [Questions] [answered]
I really want determinism for a game I'm trying to make in Godot (I want to make a lockstep implementation).

The only issue is that it's clear that 
> While Jolt itself offers deterministic simulations, Godot Jolt is not able to make such guarantees. Simulations in Godot Jolt may look deterministic, and may even happen to be deterministic, but this should not be relied upon if determinism is a hard requirement.

Is there a way I can help make this happen? I'm not necessarily proficient in c++, but I'm a software engineer with 4+ years experience in the industry and am willing to figure out what needs to be done to make this happen.

As for requirements: I don't necessarily need all platforms to be deterministic, I can limit this to Windows.

I've done some basic research on the topic of determinism. My current understanding is that things are not deterministic because the floating point math done on different machines with different OSes aren't the same. Even on the same machine, it may not be guaranteed.

**Comment by torky** [comment]
Note, a few more thoughts here,

If I could convert everything to integers, that would also be viable.

I have a feeling this may be an impossible task given the current restraints.

I've read some documentation on the main jolt github, maybe I can just custom compile some changes made to godot and jolt and be fine with that?

**Comment by mihe** [comment, marked as answer]
If you just need me to enable Jolt's deterministic simulations (for a single platform with the same binary), then nothing needs to be done. I haven't disabled it. In fact, if you run any scene multiple times with this extension you'll probably find that the outcome looks about the same every time.

What I can't give you is the strict guarantee that every line of code that runs in your game (including Godot itself) is perfectly deterministic. For that someone would need to do the work of actually testing that everything down to the bit level is exactly the same in every scenario all the time.

If you do go the route of using this for some sort of lockstepping and you can point to some specific part of the code in this extension that isn't deterministic (on a single platform with the same binary) then I can maybe address that particular issue, but I don't have any interest in labeling this extension as perfectly deterministic. It would be irresponsible without me having done the extensive testing to verify that it actually is.
